src/pybind/mgr/smb/mon_store.py

```diff
@@ -0,0 +1,233 @@
+from typing import Collection, Dict, Iterator, Optional, cast
+
+import json
+
+from .proto import EntryKey, MonCommandIssuer, Protocol, Simplified
+
+
+class MgrStoreProtocol(Protocol):
+    """A simple protocol describing the minimal per-mgr-module (mon) store interface
+    provided by the fairly giganto MgrModule class.
+    """
+
+    def get_store(self, key: str) -> Optional[str]:
+        ...
+
+    def set_store(self, key: str, val: Optional[str]) -> None:
+        ...
+
+    def get_store_prefix(self, key_prefix: str) -> Dict[str, str]:
+        ...
+
+
+def _ksplit(key: str, prefix: str = '') -> EntryKey:
+    if prefix and key.startswith(prefix):
+        plen = len(prefix)
+        key = key[plen:]
+    ek = tuple(key.split('/', 1))
+    assert len(ek) == 2
+    # the cast is needed for older mypy versions, where asserting
+    # the length doesn't narrow the type
+    return cast(EntryKey, ek)
+
+
+def _kjoin(key: EntryKey) -> str:
+    assert len(key) == 2
+    return '/'.join(key)
+
+
+class ModuleStoreEntry:
+    """A store entry for the manager module config store."""
+
+    def __init__(
+        self, module_store: 'ModuleConfigStore', key: EntryKey
+    ) -> None:
+        self._store = module_store
+        self._key = key
+        self._store_key = self._store.PREFIX + _kjoin(key)
+
+    def set(self, obj: Simplified) -> None:
+        value = json.dumps(obj)
+        self._store._mstore.set_store(self._store_key, value)
+
+    def get(self) -> Simplified:
+        value = self._store._mstore.get_store(self._store_key)
+        if value is None:
+            raise KeyError(self._key)
+        return json.loads(value)
+
+    def remove(self) -> bool:
+        return self._store.remove(self.full_key)
+
+    def exists(self) -> bool:
+        return self._key in set(self._store)
+
+    @property
+    def uri(self) -> str:
+        ns, name = self._key
+        return f'ceph-smb-resource:{ns}/{name}'
+
+    @property
+    def full_key(self) -> EntryKey:
+        return self._key
+
+
+class ModuleConfigStore:
+    """A store that serve's as a layer on top of a mgr module's key/value store.
+    Most appropriate for the smb module internal store.
+    """
```

Comment:
It's the same ultimate store, but two different APIs with different behaviors.
In MonKeyConfigStore we have full control of the "paths" and thus is good for interfacing with external things (things that are outside of the current mgr module). ModuleConfigStore is tied to mgr specific apis that prefix the path in the key-store with the name/id of the module. It also does caching.

I could have used only MonKeyConfigStore as it is the more flexible one. But I decided to do both because of the caching in the mgr module specific api.

Reply:
I updated the docstrings on these classes